src/term.rs

```diff
@@ -287,15 +295,131 @@ pub struct RecordDeps {
 /// Potential dependencies of a single field over the sibling fields in a recursive record.
 pub type FieldDeps = Option<Rc<HashSet<Ident>>>;
 
-#[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Copy, Clone)]
+/// A wrapper around f64 which makes `NaN` not representable. As opposed to floats, it is `Eq` and
+/// `Ord`.
+#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
+pub struct NumeralPriority(f64);
+
+/// Error raised when trying to convert a float with `NaN` value to a `NumeralPriority`.
+#[derive(Debug, Copy, Clone)]
+pub struct PriorityIsNaN;
+
+// The following impl are ok because `NumeralPriority(NaN)` can't be constructed.
+impl Eq for NumeralPriority {}
+impl Ord for NumeralPriority {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // Ok: NaN is forbidden
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+impl NumeralPriority {
+    pub fn zero() -> Self {
+        NumeralPriority(0.0)
+    }
+}
+
+impl TryFrom<f64> for NumeralPriority {
+    type Error = PriorityIsNaN;
+
+    fn try_from(f: f64) -> Result<Self, Self::Error> {
+        if f.is_nan() {
+            Err(PriorityIsNaN)
+        } else {
+            Ok(NumeralPriority(f))
+        }
+    }
+}
+
+impl From<NumeralPriority> for f64 {
+    fn from(n: NumeralPriority) -> Self {
+        n.0
+    }
+}
+
+impl fmt::Display for NumeralPriority {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
 pub enum MergePriority {
-    Default,
-    Normal,
+    /// The priority of default values that are overridden by everything else.
+    Bottom,
+    /// The priority by default, when no priority annotation (`default`, `force`, `priority`) is
+    /// provided.
+    ///
+    /// Act as the value `MergePriority::Numeral(0.0)` with respect to ordering and equality
+    /// testing. The only way to discriminate this variant is to pattern match on it.
+    Neutral,
+    /// A numeral priority. The inner value should never be `NaN`. Comparing a `MergePriority` with
+    /// a `NaN` value will panic.
```

Comment:
This comment isn't needed anymore, now that I've defined a proper `NumeralPriority` wrapper around floats.

```suggestion
    /// A numeral priority.
```